### PR TITLE
feat: Immutable ledger

### PR DESCRIPTION
**Pull request by deepeshgarg007**
Requires: https://github.com/frappe/frappe/pull/8255
Requires: https://github.com/frappe/frappe/pull/8290

This PR introduces the following changes in accounting and ledger generation:
- On cancellation of any document instead of deleting the GL Entry or SLE a reverse entry will be posted against that document
- Backdated entries will only be allowed after the last stock transaction's timestamp
- Since backdated entries are disabled there is no need of reposting future entries so removed logic for reposting future entries
- User will be able to filter cancelled doc entries which might be needed for audit purposes.

### CODE REVIEW

**Comment by deepeshgarg007** [comment]
Do not close

File: erpnext/patches/v12_0/update_is_cancelled_field.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+from __future__ import unicode_literals
+import frappe
+from erpnext.regional.india.setup import make_custom_fields
```
Comment (Don-Leopardo) [inline erpnext/patches/v12_0/update_is_cancelled_field.py]: Unused import? And maybe you should print the exception in te execute. None should be pass? isn't it more clear?

**Review by nabinhait — CHANGES_REQUESTED** [review]
Travis failing

File: erpnext/controllers/accounts_controller.py [hunk 2]
```diff
@@ -681,6 +681,10 @@ def set_total_advance_paid(self):
 
 		if advance:
 			advance = advance[0]
+
+			if self.doctype == "Sales Order":
+				advance['amount'] = - advance['amount']
```
Comment (nabinhait) [inline erpnext/controllers/accounts_controller.py]: What does it mean?

File: erpnext/stock/stock_balance.py [hunk 3]
```diff
@@ -241,6 +239,41 @@ def reset_serial_no_status_and_warehouse(serial_nos=None):
 			except:
 				pass
 
+def repost_all_stock_vouchers():
```
Comment (nabinhait) [inline erpnext/stock/stock_balance.py]: remove this function

File: erpnext/patches/v12_0/update_is_cancelled_field.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+		frappe.db.sql("UPDATE `tabStock Ledger Entry` SET is_cancelled = 0 where is_cancelled in ('', NULL, 'No')")
+		frappe.db.sql("UPDATE `tabSerial No` SET is_cancelled = 0 where is_cancelled in ('', NULL, 'No')")
+
+		frappe.db.sql("UPDATE `tabStock Ledger Entry` SET is_cancelled = 1 where is_cancelled = 'Yes'")
+		frappe.db.sql("UPDATE `tabSerial No` SET is_cancelled = 1 where is_cancelled = 'Yes'")
+
+		frappe.reload_doc("stock", "doctype", "stock_ledger_entry")
+		frappe.reload_doc("stock", "doctype", "serial_no")
+	except:
+		None
```
Comment (nabinhait) [inline erpnext/patches/v12_0/update_is_cancelled_field.py]: pass

**Review by nabinhait — CHANGES_REQUESTED** [review]

File: erpnext/controllers/stock_controller.py [hunk 5]
```diff
@@ -34,10 +34,6 @@ def make_gl_entries(self, gl_entries=None, repost_future_gle=True, from_repost=F
 					gl_entries = self.get_gl_entries(warehouse_account)
 				make_gl_entries(gl_entries, from_repost=from_repost)
 
-			if (repost_future_gle or self.flags.repost_future_gle):
-				items, warehouses = self.get_items_and_warehouses()
-				update_gl_entries_after(self.posting_date, self.posting_time, warehouses, items,
```
Comment (nabinhait) [inline erpnext/controllers/stock_controller.py:39]: remove update_gl_entries_after function

**Review by nabinhait — CHANGES_REQUESTED** [review]
Simplify `build` method in stock_ledger.py

**Review by nabinhait — APPROVED** [review]

**Pull request merged**